---
title: "SCN2A Mutations"
output: html_notebook
---

# SCN2A mutations from Sanders et al. (2018)
Update: 2019/09/03

### Background

*SCN2A* is a voltage-gated sodium channel gene that encodes the neuronal sodium channel *NaV1.2* and plays a critical role in action potential initiation during early neurodevelopment. The latest study demonstrated that it is loss of function mutations that in *SCN2A* that lead to autism spectrum disorders (ASD), in contrast to gain of function, which leads to infantile seizures ([Ben-Shalom 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5796785/)). 

In this tutorial, we will handle genetic data for *SCN2A* mutations identified in latest genomic studies, and explore their patterns using R basic functions. Our tutorial will utilize the summary data from [Sanders et al. (2018)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6015533/). We can obtain the list of mutations in the supplementary table 1 (Table S1).

### Aims

What we will do with this dataset

- Be familiar with a data frame format.
- Practice functions you have learnt from the Chapter 3 and 4.
- Unboxing the dataset from a scientific journal ;)

## Load dataset

We will read the txt file from the shared Dropbox link into your workspace. The file includes the first tab of the Table S1. You use `read.csv` function to read the file from the URL. Then, you create the `d` object.

```{r}
options(stringsAsFactors = F)
d = read.csv('https://www.dropbox.com/s/he7r6v4hxu0ibno/tableS1.Sanders2018.txt?dl=1', sep='\t')
```

## Explore your data

Let's explore the object you just loaded. How would you check the class of the object `d`?

```{r}
# Please type the command below
class(d)
```

It shows that the `d` object is '_________'. 

Then, let's overview this data frame. We will use `head` function to print out the first few lines.

```{r}
head(d)
```

When you execute code in a notebook chunk, an output will be visible immediately beneath the input. From this, you can see several rows and columns in the data frame. Let's look at the first column `PatientID` and check which class it is. 

```{r}
# Please type the command below
head(d$PatientID)
```

Right. It's '___________'. How about the class of the column `Effect`?

```{r}
# Please type the command below
class(d$TrueRecurrence)
```

To check the class of columns, you don't need to type an individual column. We can overview the summary of the dataset using `summary` function. Which column has the `character` class?

```{r}
# Please type the command below
summary(d)
```

## Difference between data frame and matrix

Here we will convert the data frame into a matrix, and compare which part will be different in this. To convert a data frame into a matrix, you can use the command called `as.matrix`.

```{r}
m = as.matrix(d)
```

Let's overview the matrix object. Can you tell difference with data frame?

```{r}
head(m[,'TrueRecurrence'])
```

## Subsetting and Sorting

Some patients who have the SCN2A mutation (hereafter called "SCN2A patient") often have seizures. So we want to know when the seizure occurs in development.

Let's check the class first.

```{r}
class(d$SeizureOnsetDays)
```

Why this column contains character? Let's head the first few lines.

```{r}
head(d$SeizureOnsetDays)
```

It seems that some rows contain samples who do not have seizure or unknown information. It's represented by `"."`, and also recorded in another column called `Seizures`.

```{r}
head(d$Seizures)
```

So we want to subset rows where the seizure phenotype is available.

```{r}
d1 = d[d$Seizures=='Y',]
```

Let's see how many samples have seizure phenotypes?
Then, you can ask when is the earliest days for the representation of seizure phenotype? How can we check this? The fisrt, as seen previously the `SeizureOnsetDays` column is character so we cannot apply functions for numeric. 

```{r}
head(d1$SeizureOnsetDays)
```

So we have to convert this into numeric first. 

```{r}
d1$SeizureOnsetDays2 <- as.numeric(d1$SeizureOnsetDays)
```

Hmm. There's an warning for `NA` introduction. This is because some rows do not have character that we can properly convert from character to numeric. So possible solutions are either you can bear with this in your downstream analyses or 2) convert character into an appropriate form of numeric conversion.

Then, the question is how can we find the rows with `NA`? We will ask whether the rows contains `NA` or not using `is.na` function. This will return boolean as to `NA` presence.

```{r}
head(is.na(d1$SeizureOnsetDays2))
```

See we can find some rows with `NA`. One of them is the 7th row. Let's see how it looks like.

```{r}
d1$SeizureOnsetDays[7]
```

Here you have `<` (angle bracket) in the character so it won't properly converted to numeric information. Did you find more of these cases?

```{r}
d1[is.na(d1$SeizureOnsetDays2),]$SeizureOnsetDays
```

Our `NAs` all contains `<`, which prevent converting a character into a numeric.

We would fix for downstream analyses. For example, we can convert `<365` into `365`. One function we can try is `gsub`. This replace your string into a format that you may not get `NA`. For example, 

```{r}
# gsub('pattern in your character', 'new character you want to replace', vectors for your character)
d1$SeizureOnsetDays3 <- gsub('<', '', d1$SeizureOnsetDays)
head(d1$SeizureOnsetDays3)
```

Let's convert them into numerics.

```{r}
d1$SeizureOnsetDays3 <- as.numeric(d1$SeizureOnsetDays3)
```

Did you get warning for this?
Now we can ask our initial question. When is the earliest day for having seizure?

```{r}
min(d1$SeizureOnsetDays3)
```


## Exercise

- Load the dataset using functions from the `tibble` package.
- Describe columns and set up a plan for downstream analyses.
- Explore other columns for frequency.

## Where is the genetic variants occurring? For-loops and Vectorization

The dataset includes two columns - `c.DNA` and `p.Protein` for the position of the genetic mutations in SCN2A. From this information, what can we analyze further?

1. Finding the position of the genetic mutations within SCN2A. Which information you would use?
2. Counting the recurrent mutations at the same position (in other words, the same mutations from different patients), and examine whether the patients have similar phenotype.
3. Finding the position where different consequence or amino acid change of mutations occur. Please note that "consequences" are loss-of-function (Nonsense, Frameshift) or missense. 
4. Think about the scheme for visualizing the mutation positions. 



<img src="https://ars.els-cdn.com/content/image/1-s2.0-S1525157810603546-gr2.gif" width="500">

**Figure 1. Standard mutation nomenclature** ([Ogino et al. 2007](https://www.sciencedirect.com/science/article/pii/S1525157810603546)). According to this guideline, genetic mutations can be represented for coding DNA reference sequences (`c.` prefix ) and protein-level amino acid sequences (`p.` prefix). 


```{r}
# Remove non-numeric characters from the string
gsub('[^0-9]', '', 'p.R102X')
```




